packages/concerto-core/lib/introspect/mapkeytype.js

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const { MetaModelNamespace } = require('@accordproject/concerto-metamodel');
+
+const Decorated = require('./decorated');
+const IllegalModelException = require('./illegalmodelexception');
+const ModelUtil = require('../modelutil');
+
+// Types needed for TypeScript generation.
+/* eslint-disable no-unused-vars */
+/* istanbul ignore next */
+if (global === undefined) {
+    const ModelFile = require('./modelfile');
+}
+
+/**
+ * MapKeyType defines a Key type of an MapDeclaration.
+ *
+ * @extends Decorated
+ * @see See {@link Decorated}
+ * @class
+ * @memberof module:concerto-core
+ */
+class MapKeyType extends Decorated {
+    /**
+     * Create an MapKeyType.
+     * @param {MapDeclaration} parent - The owner of this property
+     * @param {Object} ast - The AST created by the parser
+     * @throws {IllegalModelException}
+     */
+    constructor(parent, ast) {
+        super(ast);
+        this.parent = parent;
+        this.name = null;
+        this.type = null;
+        this.fqn = null;
+        this.process();
+    }
+
+    /**
+     * Process the AST and build the model
+     *
+     * @throws {IllegalModelException}
+     * @private
+     */
+    process() {
+        super.process();
+        this.name = this.ast.name;
+        this.type = this.ast.name;
+        this.fqn = ModelUtil.getFullyQualifiedName(this.parent.getModelFile().getNamespace(), this.ast.name);
```

Comment:
I'm not convinced that this is right.

Do we ever want to reference a key or value outside of the context of a MapDeclaration?

If so, the fqn should have the prefix of the parent MapDeclaration's fqn, not just it's namespace.
If not, a fqn doesn't make much sense for keys/values